pkg/autoscaling/calculation.go

```diff
@@ -52,89 +49,399 @@ var (
+	if strategy.Rules[0].StorageRule == nil || len(instances) == 0 {
+		return nil, nil
+	}
+
+	// get total storage used size and total storage capacity
+	totalStorageUsedSize, totalStorageCapacity, err := getTotalStorageInfo(rc, instances)
+	if err != nil {
+		return nil, err
+	}
+
+	// calculate storage usage
+	storageUsage := totalStorageUsedSize / totalStorageCapacity
+	storageMaxThreshold, storageMinThreshold := getStorageThresholdByComponent(strategy, TiKV)
+	storageUsageTarget := (storageMaxThreshold + storageMinThreshold) / 2
+
+	log.Debug("get storage usage information completed",
+		zap.Float64("totalStorageUsedSize", totalStorageUsedSize),
+		zap.Float64("totalStorageUsedSize", totalStorageCapacity),
+		zap.Float64("totalStorageCapacity", totalStorageCapacity),
+		zap.Float64("storageUsage", storageUsage),
+		zap.Float64("storageMaxThreshold", storageMaxThreshold),
+		zap.Float64("storageMinThreshold", storageMinThreshold))
+
+	if storageUsage > storageMaxThreshold {
+		// generate homogeneous tikv plan
+		resources := getStorageResourcesByComponent(strategy, TiKV)
+		homogeneousTiKVCount := getCountByResourceType(resources, homogeneousTiKVResourceType)
+
+		if resourceMap[homogeneousTiKVResourceType] == strategy.NodeCount || (homogeneousTiKVCount != nil && resourceMap[homogeneousTiKVResourceType] > *homogeneousTiKVCount) {
+			// homogeneous instance number reaches k8s node number or the resource limit,
+			// can not scale out homogeneous instance anymore
+			log.Warn("can not scale out homogeneous instance",
+				zap.Uint64("homogeneous instance number", resourceMap[homogeneousTiKVResourceType]),
+				zap.Uint64("k8s node number", strategy.NodeCount),
+				zap.Uint64p("resource limit", homogeneousTiKVCount),
+			)
+			return nil, nil
 		}
+
+		storageScaleSize := totalStorageUsedSize/storageUsageTarget - totalStorageCapacity
+		storeStorageSize := getStorageByResourceType(resources, homogeneousTiKVResourceType)
+		scaleOutCount := uint64(storageScaleSize)/storeStorageSize + 1
+
+		return getHomogeneousScaleOutPlans(scaleOutCount, uint64(len(instances)), strategy.NodeCount, resources, resourceMap, TiKV), nil
 	}
 
-	return plans
+	return nil, nil
 }
 
-func getPlans(rc *cluster.RaftCluster, querier Querier, strategy *Strategy, component ComponentType) []*Plan {
-	var instances []instance
-	if component == TiKV {
-		instances = filterTiKVInstances(rc)
-	} else {
-		instances = getTiDBInstances(rc.GetEtcdClient())
+func getPrometheusAddress(rc *cluster.RaftCluster) (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+
+	resp, err := rc.GetEtcdClient().Get(ctx, prometheusAddressKey)
+	if err != nil {
+		return "", err
+	}
+	if len(resp.Kvs) == 0 {
+		return "", errors.New(fmt.Sprintf("length of the response values of the key %s is 0", prometheusAddressKey))
```

Comment:
How about using `errors.Errorf` for all `errors.New(fmt.Sprintf())`?

Reply:
done